File: backend/app/schemas/department.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


class DepartmentItem(BaseModel):
    id: int
    hospital_id: int
    hospital_name: str | None = None
    code: str
    name: str
    is_active: bool
    created_at: datetime


class DepartmentCreateRequest(BaseModel):
    hospital_id: int
    code: str | None = Field(default=None, min_length=2, max_length=50)
    name: str = Field(min_length=2, max_length=100)


class DepartmentUpdateRequest(BaseModel):
    code: str | None = Field(default=None, min_length=2, max_length=50)
    name: str | None = Field(default=None, min_length=2, max_length=100)
    is_active: bool | None = None


class DepartmentAssignUsersRequest(BaseModel):
    user_ids: list[int] = Field(default_factory=list)
    mode: Literal["append", "replace"] = "append"


class ApiResponse(BaseModel):
    code: int
    message: str
    data: DepartmentItem | list[DepartmentItem] | None = None

File: frontend/src/api/department.ts
import request from './request'
import { cachedFetch, invalidateCache } from '../utils/apiCache'

interface ApiResponse<T> {
  code: number
  message: string
  data: T
}

export interface DepartmentItem {
  id: number
  hospital_id: number
  hospital_name?: string | null
  code: string
  name: string
  is_active: boolean
  created_at: string
}

export const getDepartments = async (params?: {
  active_only?: boolean
  hospital_id?: number
}) => {
  const cacheKey = `departments:${params?.active_only ?? 'all'}:${params?.hospital_id ?? 'all'}`
  return cachedFetch(cacheKey, async () => {
    const res = await request.get<ApiResponse<DepartmentItem[]>>('/departments', { params })
    return res.data.data
  })
}

export const invalidateDepartmentCache = () => invalidateCache('departments:')

export const createDepartment = async (payload: {
  hospital_id: number
  name: string
  code?: string
}) => {
  const res = await request.post<ApiResponse<DepartmentItem>>('/departments', payload)
  invalidateDepartmentCache()
  return res.data.data
}

export const updateDepartment = async (
  departmentId: number,
  payload: { name?: string; is_active?: boolean; code?: string },
) => {
  const res = await request.put<ApiResponse<DepartmentItem>>(`/departments/${departmentId}`, payload)
  invalidateDepartmentCache()
  return res.data.data
}

export const toggleDepartment = async (departmentId: number) => {
  const res = await request.delete<ApiResponse<DepartmentItem>>(`/departments/${departmentId}`)
  invalidateDepartmentCache()
  return res.data.data
}

export const assignUsersToDepartment = async (
  departmentId: number,
  payload: { user_ids: number[]; mode?: 'replace' | 'append' },
) => {
  const res = await request.put<ApiResponse<DepartmentItem>>(
    `/departments/${departmentId}/assign-users`,
    payload,
  )
  return res.data.data
}
